python2Packages.nixpart0 doesn't build

**Describe the bug**

`python2Packages.nixpart0` fails to build. 

Tested on both

* current nixpkgs (commit 01790228caaeb9ac8e16bb9ef32a925592159248)
* nixpkgs 20.03 

Impact: It is apparently used as part of nixops deploy on hetzner, so my deployment is blocked on this.

**To Reproduce**
Steps to reproduce the behavior:

`nix build -L -f '<nixpkgs>' python2Packages.nixpart0`
or 
`nix build -L -f . python2Packages.nixpart0`

in the checked out nixpkgs sources.

**Expected behavior**

The package builds without errors.

**Screenshots**

The end of the build output:

```
  /nix/store/kgp3vq8l9yb8mzghbw83kyr3f26yqvsz-bash-4.4-p23/bin/bash ../libtool  --tag=CC   --mode=link gcc  -Wall -g -O2   -o cryptsetup cryptsetup-utils_crypt.o cryptsetup-utils_loop.o cryptsetup-utils_fips.o cryptsetup-utils_tools.o cryptsetup-utils_password.o cryptsetup-cryptsetup.o ../lib/libcryptsetup.la -lpopt  
  libtool: link: gcc -Wall -g -O2 -o .libs/cryptsetup cryptsetup-utils_crypt.o cryptsetup-utils_loop.o cryptsetup-utils_fips.o cryptsetup-utils_tools.o cryptsetup-utils_password.o cryptsetup-cryptsetup.o  ../lib/.libs/libcryptsetup.so -lpopt -Wl,-rpath -Wl,/nix/store/gdyd8rvq0982axbhfabjikqlgvih7nqp-cryptsetup-1.6.3/lib
  /nix/store/3b3ighb83nhifa1v4n7855hlbdl1mhf9-binutils-2.31.1/bin/ld: cryptsetup-utils_loop.o: in function `_sysfs_backing_file':
  /build/cryptsetup-1.6.3/src/../lib/utils_loop.c:192: undefined reference to `minor'
  /nix/store/3b3ighb83nhifa1v4n7855hlbdl1mhf9-binutils-2.31.1/bin/ld: /build/cryptsetup-1.6.3/src/../lib/utils_loop.c:192: undefined reference to `major'
  /nix/store/3b3ighb83nhifa1v4n7855hlbdl1mhf9-binutils-2.31.1/bin/ld: cryptsetup-utils_loop.o: in function `crypt_loop_device':
  /build/cryptsetup-1.6.3/src/../lib/utils_loop.c:222: undefined reference to `major'
  /nix/store/3b3ighb83nhifa1v4n7855hlbdl1mhf9-binutils-2.31.1/bin/ld: ../lib/.libs/libcryptsetup.so: undefined reference to `makedev'
  collect2: error: ld returned 1 exit status
  make[2]: *** [Makefile:613: cryptsetup] Error 1
  make[2]: Leaving directory '/build/cryptsetup-1.6.3/src'
  make[1]: *** [Makefile:451: all-recursive] Error 1
  make[1]: Leaving directory '/build/cryptsetup-1.6.3'
  make: *** [Makefile:382: all] Error 2
cannot build derivation '/nix/store/45lhiz89i33x1w1sy8i1q7qk5xrrahn8-blivet-0.17-1.drv': 1 dependencies couldn't be built
cannot build derivation '/nix/store/c36z6ivydb5wpp6zzban6vd401mwk1yg-nixpart-0.4.1.drv': 1 dependencies couldn't be built
error: build of '/nix/store/c36z6ivydb5wpp6zzban6vd401mwk1yg-nixpart-0.4.1.drv' failed
```
[Here is the full build output](https://gist.github.com/kolloch/6143f75e5b27b190a848728a2385ff72).

**Additional context**

I found [ancient cryptsetup mailing list references](https://www.saout.de/pipermail/dm-crypt/2016-April/005172.html) to this build problem. I tried to patch in sysmacro includes to fix the problem without avail.

The URL used in the cryptsetup source, doesn't work anymore. I used the gitlab sources but they don't include the configure script yet, so I needed additional buildInputs to run "autogen.sh".

These are my attempts:
https://github.com/kolloch/nixpkgs/commit/60486dbcae0aedf4f56f5e11735e85e40610f187

**Notify maintainers**
@aszlig

**Metadata**
```
 - system: `"x86_64-linux"`
 - host os: `Linux 5.4.0-26-generic, Ubuntu, 20.04 LTS (Focal Fossa)`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.3.4`
 - channels(root): `""`
 - channels(peter): `"home-manager, nixpkgs-20.03.1619.ab3adfe1c76"`
 - nixpkgs: `/home/peter/.nix-defexpr/channels/nixpkgs`
```

Maintainer information:
```yaml
# a list of nixpkgs attributes affected by the problem
attribute: [ "python2Packages.nixpart0" ]
# a list of nixos modules affected by the problem
module:
```

--- a/pkgs/tools/filesystems/nixpart/0.4/cryptsetup.nix
+++ b/pkgs/tools/filesystems/nixpart/0.4/cryptsetup.nix
@@ -1,4 +1,4 @@
-{ stdenv, fetchurl, lvm2, libgcrypt, libuuid, pkgconfig, popt
+{ stdenv, fetchurl, fetchpatch, lvm2, libgcrypt, libuuid, pkgconfig, popt
 , enablePython ? true, python ? null
 }:
 
@@ -11,6 +11,15 @@ stdenv.mkDerivation rec {
     url = "http://cryptsetup.googlecode.com/files/${name}.tar.bz2";
     sha256 = "1n1qk5chyjspbiianrdb55fhb4wl0vfyqz2br05vfb24v4qlgbx2";
   };
+
+  patches = [
+    # Fix build with glibc >= 2.28
+    # https://github.com/NixOS/nixpkgs/issues/86403
+    (fetchpatch {
+      url = "https://gitweb.gentoo.org/repo/gentoo.git/plain/sys-fs/cryptsetup/files/cryptsetup-1.7.1-sysmacros.patch?id=d72316f97ebcc7fe622b21574442a9ac59b9115f";
+      sha256 = "0xbhazgl44bimqhcrhajk016w9wi7bkrgwyfq13xmrvyrllqvgdx";
+    })
+  ];
 
   configureFlags = [ "--enable-cryptsetup-reencrypt" ]
                 ++ stdenv.lib.optional enablePython "--enable-python";

--- a/pkgs/tools/filesystems/nixpart/0.4/default.nix
+++ b/pkgs/tools/filesystems/nixpart/0.4/default.nix
@@ -27,7 +27,7 @@ let
 
   lvm2 = import ./lvm2.nix {
     inherit stdenv fetchurl;
-    inherit (pkgs) pkgconfig utillinux systemd coreutils;
+    inherit (pkgs) fetchpatch pkgconfig utillinux systemd coreutils;
   };
 
   multipath_tools = import ./multipath-tools.nix {

--- a/pkgs/tools/filesystems/nixpart/0.4/default.nix
+++ b/pkgs/tools/filesystems/nixpart/0.4/default.nix
@@ -18,7 +18,7 @@ let
 
   cryptsetup = import ./cryptsetup.nix {
     inherit stdenv fetchurl python;
-    inherit (pkgs) pkgconfig libgcrypt libuuid popt lvm2;
+    inherit (pkgs) fetchpatch pkgconfig libgcrypt libuuid popt lvm2;
   };
 
   dmraid = import ./dmraid.nix {

--- a/pkgs/tools/filesystems/nixpart/0.4/default.nix
+++ b/pkgs/tools/filesystems/nixpart/0.4/default.nix
@@ -32,7 +32,7 @@ let
 
   multipath_tools = import ./multipath-tools.nix {
     inherit stdenv fetchurl lvm2;
-    inherit (pkgs) readline systemd libaio gzip;
+    inherit (pkgs) fetchpatch readline systemd libaio gzip;
   };
 
   parted = import ./parted.nix {

--- a/pkgs/tools/filesystems/nixpart/0.4/default.nix
+++ b/pkgs/tools/filesystems/nixpart/0.4/default.nix
@@ -37,7 +37,7 @@ let
 
   parted = import ./parted.nix {
     inherit stdenv fetchurl;
-    inherit (pkgs) utillinux readline libuuid gettext check lvm2;
+    inherit (pkgs) fetchpatch utillinux readline libuuid gettext check lvm2;
   };
 
   pyblock = import ./pyblock.nix {

--- a/pkgs/tools/filesystems/nixpart/0.4/lvm2.nix
+++ b/pkgs/tools/filesystems/nixpart/0.4/lvm2.nix
@@ -1,4 +1,4 @@
-{ stdenv, fetchurl, pkgconfig, systemd, utillinux, coreutils }:
+{ stdenv, fetchurl, fetchpatch, pkgconfig, systemd, utillinux, coreutils }:
 
 let
   v = "2.02.106";
@@ -11,6 +11,18 @@ stdenv.mkDerivation {
     url = "ftp://sources.redhat.com/pub/lvm2/releases/LVM2.${v}.tgz";
     sha256 = "0nr833bl0q4zq52drjxmmpf7bs6kqxwa5kahwwxm9411khkxz0vc";
   };
+
+  patches = [
+    # Fix build with glibc >= 2.28
+    # https://github.com/NixOS/nixpkgs/issues/86403
+    (fetchpatch {
+      url = "https://github.com/lvmteam/lvm2/commit/92d5a8441007f578e000b492cecf67d6b8a87405.patch";
+      sha256 = "1yqd6jng0b370k53vks1shg57yhfyribhpmv19km5zsjqf0qqx2d";
+      excludes = [
+        "libdm/libdm-stats.c"
+      ];
+    })
+  ];
 
   configureFlags = [
     "--disable-readline"

--- a/pkgs/tools/filesystems/nixpart/0.4/multipath-tools.nix
+++ b/pkgs/tools/filesystems/nixpart/0.4/multipath-tools.nix
@@ -38,6 +38,11 @@ stdenv.mkDerivation rec {
 
       substituteInPlace libmultipath/defaults.h --replace /lib/udev/scsi_id ${systemd.lib}/lib/udev/scsi_id
       substituteInPlace libmultipath/hwtable.c --replace /lib/udev/scsi_id ${systemd.lib}/lib/udev/scsi_id
+
+      sed -i -re '
+         s,^( *#define +DEFAULT_MULTIPATHDIR\>).*,\1 "'"$out/lib/multipath"'",
+      ' libmultipath/defaults.h
+
     '';
 
   meta = {

--- a/pkgs/tools/filesystems/nixpart/0.4/multipath-tools.nix
+++ b/pkgs/tools/filesystems/nixpart/0.4/multipath-tools.nix
@@ -1,6 +1,6 @@
 # FIXME: unify with pkgs/os-specific/linux/multipath-tools/default.nix.
 
-{ stdenv, fetchurl, lvm2, libaio, gzip, readline, systemd }:
+{ stdenv, fetchurl, fetchpatch, lvm2, libaio, gzip, readline, systemd }:
 
 stdenv.mkDerivation rec {
   name = "multipath-tools-0.4.9";
@@ -9,6 +9,18 @@ stdenv.mkDerivation rec {
     url = "http://christophe.varoqui.free.fr/multipath-tools/${name}.tar.bz2";
     sha256 = "04n7kazp1zrlqfza32phmqla0xkcq4zwn176qff5ida4a60whi4d";
   };
+
+  patches = [
+    # Fix build with glibc >= 2.28
+    # https://github.com/NixOS/nixpkgs/issues/86403
+    (fetchpatch {
+      url = "https://gitweb.gentoo.org/repo/gentoo.git/plain/sys-fs/multipath-tools/files/multipath-tools-0.6.4-sysmacros.patch?id=eb22b954c177b5c1e2b6ed5c7cdd02f40f40d757";
+      sha256 = "1an0cgmz7g03c4qjimhpm9fcf2iswws18lwqxi688k87qm3xb5qd";
+      excludes = [
+        "libmultipath/util.c"
+      ];
+    })
+  ];
 
   sourceRoot = ".";
 

--- a/pkgs/tools/filesystems/nixpart/0.4/parted.nix
+++ b/pkgs/tools/filesystems/nixpart/0.4/parted.nix
@@ -1,4 +1,4 @@
-{ stdenv, fetchurl, lvm2, libuuid, gettext, readline
+{ stdenv, fetchurl, fetchpatch, lvm2, libuuid, gettext, readline
 , utillinux, check, enableStatic ? false }:
 
 stdenv.mkDerivation rec {
@@ -8,6 +8,15 @@ stdenv.mkDerivation rec {
     url = "mirror://gnu/parted/${name}.tar.xz";
     sha256 = "05fa4m1bky9d13hqv91jlnngzlyn7y4rnnyq6d86w0dg3vww372y";
   };
+
+  patches = [
+    # Fix build with glibc >= 2.28
+    # https://github.com/NixOS/nixpkgs/issues/86403
+    (fetchpatch {
+      url = "https://gitweb.gentoo.org/repo/gentoo.git/plain/sys-block/parted/files/parted-3.2-sysmacros.patch?id=8e2414f551c14166f259f9a25a594aec7a5b9ea0";
+      sha256 = "0fdgifjbri7n28hv74zksac05gw72p2czzvyar0jp62b9dnql3mp";
+    })
+  ];
 
   buildInputs = [ libuuid ]
     ++ stdenv.lib.optional (readline != null) readline

--- a/pkgs/tools/filesystems/nixpart/0.4/pyblock-sysmacros.h.patch
+++ b/pkgs/tools/filesystems/nixpart/0.4/pyblock-sysmacros.h.patch
@@ -1,0 +1,12 @@
+diff --git a/dm.c b/dm.c
+index 5daa0e5..d5b84c8 100644
+--- a/dm.c
++++ b/dm.c
+@@ -19,6 +19,7 @@
+ #define _GNU_SOURCE
+ #include <sys/stat.h>
+ #include <sys/types.h>
++#include <sys/sysmacros.h>
+ #include <unistd.h>
+ #include <fcntl.h>
+ #include <stdarg.h>

--- a/pkgs/tools/filesystems/nixpart/0.4/pyblock.nix
+++ b/pkgs/tools/filesystems/nixpart/0.4/pyblock.nix
@@ -11,6 +11,12 @@ stdenv.mkDerivation rec {
     sha256 = "f6cef88969300a6564498557eeea1d8da58acceae238077852ff261a2cb1d815";
   };
 
+  patches = [
+    # Fix build with glibc >= 2.28
+    # https://github.com/NixOS/nixpkgs/issues/86403
+    ./pyblock-sysmacros.h.patch
+  ];
+
   postPatch = ''
     sed -i -e 's|/usr/include/python|${python}/include/python|' \
            -e 's/-Werror *//' -e 's|/usr/|'"$out"'/|' Makefile

